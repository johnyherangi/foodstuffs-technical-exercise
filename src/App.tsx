import { useState } from "react"
import "./App.css"
import { NumberInput } from "./components/NumberInput"
import { SelectInput } from "./components/SelectInput"
import { SubmitButton } from "./components/SubmitButton"
import { useCalculateTotal } from "./hooks/useCalculateTotal"
import { stateCodes, type State } from "./lib/state"

function App() {
  const { total, calculateTotal } = useCalculateTotal()
  const [items, setItems] = useState<number>()
  const [price, setPrice] = useState<number>()
  const [state, setState] = useState<State>()

  return (
    <>
      <h1>Retail Calculator</h1>
      <form action={calculateTotal} className="flex flex-col gap-4 mt-8">
        <div className="flex flex-col items-start">
          <label htmlFor="items">Number of items</label>
          <NumberInput
            id="items"
            name="items"
            min="0"
            value={items}
            onChange={(e) => setItems(Number(e.target.value))}
          />
        </div>
        <div className="flex flex-col items-start">
          <label htmlFor="price" className="text-nowrap">
            Price per item ($)
          </label>
          <NumberInput
            id="price"
            name="price"
            min="0"
            step="any"
            value={price}
            onChange={(e) => setPrice(Number(e.target.value))}
          />
        </div>
        <SelectInput
          id="state"
          name="state"
          options={stateCodes}
          placeholder="State code"
          className="w-full"
          value={state}
          onChange={setState}
        />
        <SubmitButton disabled={!items || !price || !state} className="mt-4">
          Calculate total
        </SubmitButton>
      </form>
      <div className="flex justify-between mt-8">
        <label htmlFor="total">Total</label>
        <output id="total">{total}</output>
      </div>
    </>
  )
}

export default App
