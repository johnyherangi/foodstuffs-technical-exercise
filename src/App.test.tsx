import App from "./App"
import { render, screen } from "@testing-library/react"
import { describe, it, expect, beforeAll, vi } from "vitest"
import userEvent from "@testing-library/user-event"

describe("App", () => {
  // https://stackoverflow.com/questions/79790413/how-to-avoid-target-haspointercapture-is-not-a-function-when-testing-radix-ui-s
  beforeAll(() => {
    if (!Element.prototype.hasPointerCapture) {
      Element.prototype.hasPointerCapture = vi.fn().mockReturnValue(false)
    }
    if (!Element.prototype.setPointerCapture) {
      Element.prototype.setPointerCapture = vi.fn()
    }
    if (!Element.prototype.releasePointerCapture) {
      Element.prototype.releasePointerCapture = vi.fn()
    }
    if (!Element.prototype.scrollIntoView) {
      Element.prototype.scrollIntoView = vi.fn()
    }
  })

  it("renders", () => {
    const { asFragment } = render(<App />)

    expect(asFragment()).toMatchSnapshot()
  })

  it("has accessible elements", () => {
    render(<App />)

    expect(
      screen.getByRole("spinbutton", { name: "Number of items" })
    ).toBeInTheDocument()
    expect(
      screen.getByRole("spinbutton", { name: "Price per item ($)" })
    ).toBeInTheDocument()
    expect(screen.getByLabelText("State code")).toBeInTheDocument()
    expect(
      screen.getByRole("button", { name: "Calculate total" })
    ).toBeInTheDocument()
  })

  it.each([
    {
      items: 10,
      price: 100,
      state: "AUK",
      expectedTotal: "$1,068.50",
    },
  ])(
    "calculates the correct total (items=$items, price=$price, state=$state)",
    async ({ items, price, state, expectedTotal }) => {
      render(<App />)

      const submitButton = screen.getByRole("button", {
        name: "Calculate total",
      })

      expect(submitButton).toBeDisabled()

      const user = userEvent.setup()

      const itemCount = screen.getByRole("spinbutton", {
        name: "Number of items",
      })
      await user.type(itemCount, String(items))

      const pricePerItem = screen.getByRole("spinbutton", {
        name: "Price per item ($)",
      })
      await user.type(pricePerItem, String(price))

      const stateCodeSelect = screen.getByLabelText("State code")
      await user.click(stateCodeSelect)
      const stateCode = screen.getByLabelText(state)
      await user.click(stateCode)

      expect(submitButton).toBeEnabled()
      await user.click(submitButton)

      expect(
        await screen.findByRole("status", { name: "Total" })
      ).toHaveTextContent(`${expectedTotal}`)
    }
  )
})
