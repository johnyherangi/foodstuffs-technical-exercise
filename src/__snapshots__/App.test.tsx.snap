// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders 1`] = `
<DocumentFragment>
  <h1>
    Retail Calculator
  </h1>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
    class="flex flex-col gap-4 mt-8"
  >
    <div
      class="flex flex-col items-start"
    >
      <label
        for="items"
      >
        Number of items
      </label>
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
        data-slot="input"
        id="items"
        min="0"
        name="items"
        type="number"
      />
    </div>
    <div
      class="flex flex-col items-start"
    >
      <label
        class="text-nowrap"
        for="price"
      >
        Price per item ($)
      </label>
      <input
        class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
        data-slot="input"
        id="price"
        min="0"
        name="price"
        step="any"
        type="number"
      />
    </div>
    <button
      aria-autocomplete="none"
      aria-controls="radix-_r_0_"
      aria-expanded="false"
      aria-label="State code"
      class="border-input data-placeholder:text-muted-foreground [&_svg:not([class*='text-'])]:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 dark:hover:bg-input/50 flex items-center justify-between gap-2 rounded-md border bg-transparent px-3 py-2 text-sm whitespace-nowrap shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 data-[size=default]:h-9 data-[size=sm]:h-8 *:data-[slot=select-value]:line-clamp-1 *:data-[slot=select-value]:flex *:data-[slot=select-value]:items-center *:data-[slot=select-value]:gap-2 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 w-full"
      data-placeholder=""
      data-size="default"
      data-slot="select-trigger"
      data-state="closed"
      dir="ltr"
      id="state"
      role="combobox"
      type="button"
    >
      <span
        data-slot="select-value"
        style="pointer-events: none;"
      >
        State code
      </span>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down size-4 opacity-50"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
    <select
      aria-hidden="true"
      name="state"
      style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="-1"
    >
      <option
        value=""
      />
      <option
        value="AUK"
      >
        AUK
      </option>
      <option
        value="WLG"
      >
        WLG
      </option>
      <option
        value="WAI"
      >
        WAI
      </option>
      <option
        value="CHC"
      >
        CHC
      </option>
      <option
        value="TAS"
      >
        TAS
      </option>
    </select>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-secondary-foreground hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 mt-4"
      data-slot="button"
      type="submit"
    >
      Calculate total
    </button>
  </form>
  <div
    class="flex justify-between mt-8"
  >
    <label
      for="total"
    >
      Total
    </label>
    <output
      id="total"
    />
  </div>
</DocumentFragment>
`;
